CRM/Core/I18n/Form.php

```diff
@@ -61,13 +61,36 @@ public function buildQuickForm() {
     $dao->query($query, FALSE);
     $dao->fetch();
 
-    // we want TEXTAREAs for long fields and INPUTs for short ones
-    $this->_structure[$table][$field] == 'text' ? $type = 'textarea' : $type = 'text';
+    // get html type and attributes for this field
+    $widgets = CRM_Core_I18n_SchemaStructure::widgets();
+    $widget = $widgets[$table][$field];
 
+    // attributes
+    $attributes = array('css' => '');
+    if (isset($widget['rows'])) {
+      $attributes['rows'] = $widget['rows'];
+    }
+    if (isset($widget['cols'])) {
+      $attributes['cols'] = $widget['cols'];
+    }
+    // FIXME: should have this
+    $required = !empty($widget['required']);
+    
     $languages = CRM_Core_I18n::languages(TRUE);
     foreach ($this->_locales as $locale) {
-      $this->addElement($type, "{$field}_{$locale}", $languages[$locale], array('class' => 'huge huge12' . ($locale == $tsLocale ? ' default-lang' : '')));
-      $this->_defaults["{$field}_{$locale}"] = $dao->$locale;
+      $name = "{$field}_{$locale}";
+      if ($locale == $tsLocale) {
+        $attributes['class'] .= ' default-lang';
+      }
+      if ($widget['type'] == 'RichTextEditor') {
+        $attributes['click_wysiwyg'] = TRUE;
+        $this->addWysiwyg($name, $languages[$locale], $attributes, $required);
+      }
+      else {
+        $this->add($widget['type'], $name, $languages[$locale], $attributes, $required);
+      }
+
+      $this->_defaults[$name] = $dao->$locale;
```

Comment:
Sorry, this is a bit a moving target I know, but some work we're doing in #5687 changes the syntax for creating a wysiwyg from `$this->addWysiwyg(...)` to `$this->add('wysiwyg'...)`.
And the 'click_wysiwyg' property has been changed to `class="collapsed"`